### PR TITLE
Further general improvements

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

This PR adds further general improvements for our code base.

Highlights:
- Added many additional warning flags that are useful and are neither included by `-Wall` and `-Wextra`
- Declare many `char*` which should not be changeable in the corresponding subroutines as `const char*` to have the compiler enforce their immutability.
- Enforce strict prototyping for FTL routines. This ensures that functions have to be *either* made explicitly global (by adding a prototype to the suitable header file), or they have to be made explicitly local (by adding the `static` keyword). Furthermore, strict prototyping requires that functions with zero arguments need to be written explicitly as `function(void)`. Non-strict prototype such as `function()` are permitted by regular `C`, however they in fact allow an arbitrary amount of parameters (they are a function declaration and not a prototype).
- As before, we split the warnings into a `gcc` version 8 and higher and a CI-compatible section which is automatically selected. This will be helpful for users with sufficiently recent `gcc` to get better hints whereas compilation is still possible on the CI.

Additional note: There are changes to `dnsmasq` code in this PR which are necessary to be able to compile *FTL*DNS with the added warning flags. The changes are meaningful and will be submitted to `dnsmasq` separately to be already part of the next release of `dnsmasq`.

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: grep.c [hunk 1]
```diff
@@ -45,25 +45,6 @@ int countlines(const char* fname)
 	return lines;
 }
 
-int readnumberfromfile(const char* fname)
```
Comment (DL6ER) [inline grep.c]: This function is not used anywhere

File: log.c [hunk 2]
```diff
@@ -63,7 +63,7 @@ void get_timestr(char *timestring)
 
 void __attribute__ ((format (gnu_printf, 1, 2))) logg(const char *format, ...)
 {
-	char timestring[32] = "";
+	char timestring[84] = "";
```
Comment (DL6ER) [inline log.c]: This change ensures that the string will be large enough to hold anything that might be produced by
https://github.com/pi-hole/FTL/pull/536/files#diff-f8cdfce5cea83a80a9847ba558991869R61

File: main.c [hunk 3]
```diff
@@ -14,9 +14,6 @@ char * username;
 bool needGC = false;
 bool needDBGC = false;
 
-// Prototype
-int main_dnsmasq(int argc, char **argv);
```
Comment (DL6ER) [inline main.c]: Redundant prototype, already contained in `routines.h`

File: routines.h [hunk 4]
```diff
@@ -131,7 +132,7 @@ void newOverTimeClient(int clientID);
  * Add a new overTime slot to each overTime client shared memory block.
  * This also updates `overTimeClientData`.
  */
-void addOverTimeClientSlot();
+void addOverTimeClientSlot(void);
```
Comment (DL6ER) [inline routines.h]: Changed function declaration to a full prototype

**Comment by Mcat12** [comment]
> Furthermore, strict prototyping requires that functions with zero arguments need to be written explicitly as function(void). Non-strict prototype such as function() are permitted by regular C, however they in fact allow an arbitrary amount of parameters (they are a function declaration and not a prototype).

Is this not caught by the compiler at least as a warning?

File: Makefile [hunk 5]
```diff
@@ -70,6 +70,28 @@ else
...
+# -Wbad-function-cast: Warn when a function call is cast to a non-matching type
+# -Wwrite-strings: When compiling C, give string constants the type "const char[length]" so that copying the address of one into a non-"const" "char *" pointer produces a warning
+# -Wparentheses: Warn if parentheses are omitted in certain contexts, such as when there is an assignment in a context where a truth value is expected, or when operators are nested whose precedence people often get confused about
+# -Wlogical-op: Warn about suspicious uses of logical operators in expressions
+# -Wstrict-prototypes: Warn if a function is declared or defined without specifying the argument types
+# -Wmissing-prototypes: Warn if a global function is defined without a previous prototype declaration
+# -Wredundant-decls: Warn if anything is declared more than once in the same scope
+# -Winline: Warn if a function that is declared as inline cannot be inlined
+EXTRAWARN2=-Wfloat-equal -Wunsafe-loop-optimizations -funsafe-loop-optimizations -Wbad-function-cast -Wwrite-strings -Wparentheses -Wlogical-op -Wstrict-prototypes -Wmissing-prototypes -Wredundant-decls -Winline $(EXTRAWARNGCC8)
```
Comment (Mcat12) [inline Makefile]: Can this be combined with `EXTRAWARN`?

**Comment by DL6ER** [comment]
> Is this not caught by the compiler at least as a warning?

It is now enforced by the addition of `-Wstrict-prototypes -Wmissing-prototypes`. These warnings are neither enabled by default nor included in `-Wall -Wextra` in `gcc`.

File: args.c [hunk 6]
```diff
@@ -24,8 +24,8 @@ void parse_args(int argc, char* argv[])
 	// Regardless of any arguments, we always pass "-k" (nofork) to dnsmasq
 	argc_dnsmasq = 2;
 	argv_dnsmasq = calloc(argc_dnsmasq, sizeof(char*));
-	argv_dnsmasq[0] = "";
-	argv_dnsmasq[1] = "-k";
+	argv_dnsmasq[0] = (char*)"";
+	argv_dnsmasq[1] = (char*)"-k";
```
Comment (Mcat12) [inline args.c]: Why does this cast need to exist? What other type could a string literal be?

Comment (DL6ER) [reply args.c]: `""` is `const char[1]`. However, `argv_dnsmasq` is `char*`.

```c++
argv_dnsmasq[0] = "";
```
results in the compilation error
```
error: assignment discards ‘const’ qualifier from pointer target type
```

File: Makefile (hunk 5 above)
Comment (DL6ER) [reply Makefile]: Yes, the only reason for separating was that `EXTRAWARN` is already quite long...

File: args.c (hunk 6 above)
Comment (Mcat12) [reply args.c]: Can we change `argv_dnsmasq` to `const char**` then?

File: Makefile (hunk 5 above)
Comment (Mcat12) [reply Makefile]: Makefiles support splitting lines:
https://www.gnu.org/software/make/manual/html_node/Splitting-Lines.html

**Pull request merged**